File: 数组/螺旋矩阵.go
package 数组

//递归 顺时针打印矩阵
func spiralOrder(matrix [][]int) []int {
	var res []int
	m := len(matrix)
	n := len(matrix[0])
	visited := make([][]bool, m) //visit[i][j]为0表示矩阵中的i，j位置没有访问过
	for i := 0; i < m; i++ {
		visited[i] = make([]bool, n)
	}
	visited[0][0] = true
	var dfs func(int, int)
	dfs = func(i, j int) { //现在走到死路i，j了 下一步往哪走？
		if j+1 < n && !visited[i][j+1] { //右
			var col int
			for col = j + 1; col < n && !visited[i][col]; col++ {
				res = append(res, matrix[i][col])
				visited[i][col] = true
			}
			col--
			dfs(i, col)
		}

		if i+1 < m && !visited[i+1][j] { //down
			var row int
			for row = i + 1; row < m && !visited[row][j]; row++ {
				res = append(res, matrix[row][j])
				visited[row][j] = true
			}
			row--
			dfs(row, j)
		}

		if j-1 >= 0 && !visited[i][j-1] { //left
			var col int
			for col = j - 1; col >= 0 && !visited[i][col]; col-- {
				res = append(res, matrix[i][col])
				visited[i][col] = true
			}
			col++
			dfs(i, col)
		}

		if i-1 >= 0 && !visited[i-1][j] { //up
			var row int
			for row = i - 1; row >= 0 && !visited[row][j]; row-- {
				res = append(res, matrix[row][j])
				visited[row][j] = true
			}
			row++
			dfs(row, j)
		}
	}
	res = append(res, matrix[0][0])
	dfs(0, 0)
	return res
}

//func InitVisit(m, n int) [][]bool {
//	visit := make([][]bool, m)
//	for i := 0; i < len(visit); i++ {
//		visit[i] = make([]bool, n)
//	}
//	return visit
//}

func main() {
	mat := [][]int{}
	mat = append(mat, []int{1, 2, 3})
	mat = append(mat, []int{4, 5, 6})
	mat = append(mat, []int{7, 8, 9})

}

File: 模拟/螺旋矩阵.java
package 模拟;

import java.util.ArrayList;
import java.util.List;

public class 螺旋矩阵 {
    List<Integer> res;
    int[][] matrix;
    int m;
    int n;

    public List<Integer> spiralOrder(int[][] matrix) {
        res  = new ArrayList<Integer>();
        res.add(matrix[0][0]);
        this.matrix = matrix;
        matrix[0][0] = 114514;
        m = matrix.length;
        n = matrix[0].length;
        next(0, 0);
        return res;
    }

    //我当前在 row col 位置 ，接下来怎么走交给你了
    public void next(int row, int col) {
        if (col + 1 < n && matrix[row][col + 1] != 114514) {
            int j = col + 1;
            for (; j < n && matrix[row][j] != 114514; j++) {
                res.add(matrix[row][j]);
                matrix[row][j] = 114514;
            }
            next(row, j - 1);
        }
        if (row + 1 < m && matrix[row + 1][col] != 114514) {
            int i = row + 1;
            for (; i < m && matrix[i][col] != 114514; i++) {
                res.add(matrix[i][col]);
                matrix[i][col] = 114514;
            }
            next(i - 1, col);
        }
        if (col - 1 >= 0 && matrix[row][col - 1] != 114514) {
            int j = col - 1;
            for (; j >= 0 && matrix[row][j] != 114514; j--) {
                res.add(matrix[row][j]);
                matrix[row][j] = 114514;
            }
            next(row, j + 1);
        }
        if (row - 1 >= 0 && matrix[row - 1][col] != 114514) {
            int i = row - 1;
            for (; i >= 0 && matrix[i][col] != 114514; i--) {
                res.add(matrix[i][col]);
                matrix[i][col] = 114514;
            }
            next(i + 1, col);
        }
    }
}
